### DISCUSSION TITLE
Improving Tagging and Release process

### DISCUSSION THREAD

**Discussion by fjtrujy** [pspdev/pspdev#64] [Ideas] [answered]
After having some discussions started here:
https://discord.com/channels/479828644970364928/693948645145575434/1232366213779427440

Most of the toolchain developers have agreed that improving our Tagging and Release process could help end developers to identify breaking changes and improvements.

Please continue this discussion here and let's try to define a proper plan together.

Cheers.

**Comment by diamant3** [comment]
> [!Note] 
> This discussion is related from https://github.com/pspdev/pspdev/discussions/60

**Comment by sharkwouter** [comment]
I think we'll need at least the following:
- A file with a list of which commit hash for each repo was used while building.
- We should fix the automated release system in the repos. Especially PSPDEV.
- Set versions automatically.
- Enabled automated scheduled releases.

I think things we should discuss here would be:
- Do we only do releases from the PSPDEV repo or do the other repos get tagged as well? If so, how?
- What kind of version names do we use? Semantic or date based?
- How would we add commit numbers to the build? Does PSPDEV do this or each repo by itself? Perhaps both?

There is probably more, let me know.

**Comment by fjtrujy** [comment]
There are several ways of doing this.... however I think that the easiest one is to use ["repo"](https://gerrit.googlesource.com/git-repo) in `pspdev/pspdev` repo.

**Comment by sharkwouter** [comment]
Yeah, agreed. Using pspdev/pspdev to do releases from makes sense.

**Comment by sharkwouter** [comment]
I've created the following PRs to clean up the tagging up system that we currently have that just isn't working. With these changes we can just tag pspdev and be done with it. We would still like to introduce tracking the version of the underlying repos that were used like discussed here, but that I haven't gotten to yet.

The PRs:

- https://github.com/pspdev/psptoolchain-allegrex/pull/38
- https://github.com/pspdev/psptoolchain-extra/pull/20
- https://github.com/pspdev/psptoolchain/pull/185
- https://github.com/pspdev/pspsdk/pull/208
- https://github.com/pspdev/psp-packages/pull/163
- https://github.com/pspdev/pspdev/pull/74

**Comment by sharkwouter** [comment]
Okay, I think I found out how I can add some basic versioning information, but I'm still experimenting with it.

Basically, I would like to add the following line to each repo:
```
git log -1 --format="repo-name %H %cs" >> $PSPDEV/versions.txt
```
Which results in a line like this:
```
psptoolchain-allegrex 3d6b4904609f921d493d5d5884e511c6d83be1e9 2024-06-09
```
So, name, commit hash, commit date.

For this change I would need to add `context: .` to all docker builds in the workflows, otherwise the `.git` directory is not available in the container. I'll also have to make sure that directory does not end up in the container. When this is done, we'll have a line like shown before for each repo where we can see how old the build is and how long ago it was made and what the exact commit is we are dealing with in case we want to debug an issue.

**Comment by sharkwouter** [comment]
I've created some PRs to add lines to the build information file:
- https://github.com/pspdev/psptoolchain-allegrex/pull/39
- https://github.com/pspdev/psptoolchain-extra/pull/22
- https://github.com/pspdev/psptoolchain/pull/187
- https://github.com/pspdev/pspdev/pull/76
- https://github.com/pspdev/pspsdk/pull/210

~~I'm still working on figuring out how to do this for pspsdk.~~

Do we even care about psp-packages? Technically users can just use pacman to update.

**Reply by sharkwouter** [reply]
I added a PR for pspsdk as well now. I've updated this comment.

**Comment by diamant3** [comment]
Hello guys, could we add the commit title too? Much better if we can see briefly the purpose of the commit

**Reply by sharkwouter** [reply]
I'm open for this. How would you propose that we would format the file with the commit title? Could you type out an example?

**Reply by fjtrujy** [reply]
`git log -1 --format="repo-name %H %cs %s"`
Will put the name of the commit at the end

**Reply by sharkwouter** [reply]
Will add it.

**Reply by diamant3** [reply]
Thanks guys!

**Reply by sharkwouter** [reply]
I've now added the commit title to all outstanding PRs.

**Comment by fjtrujy** [comment]
@sharkwouter I think that would be really useful to expand the information provided.... so for instance in `psptoolchain-allegrex` we just have the information of the current state of the `psptoolchain-allegrex`, however, if we apply patches to `binutils`/`gcc`/`newlib` without incrementing the version, those changes will never be reflected in the `build.txt`.
This in theory will never happen in `psp-packages` as we in theory must increment the `pkgrel`

I think that:
- binutils
- GCC
- newlib
- pthread-embedded

They should be too in the `build.txt` as they are super critical stuff in our toolchain.

**Reply by sharkwouter** [reply]
Agreed, I would like to add those.

**Comment by sharkwouter** [comment]
I added another 2 PRs. One to fix a mistake. One to add psp-pacman.

This one fixes a backup file being created that ends with `=`:

This one add psp-pacman: https://github.com/pspdev/psp-pacman/pull/32

I'll be creating more for other tools.

**Comment by sharkwouter** [comment]
I was discussing with @wally4000 about the way we do releases. He agreed that automatic monthly releases could work, but he had the idea that we might want to have a release policy. I was thinking we could have something along the following lines:

A monthly release happens automatically at the beginning of the month
If a major change is in the release, please update the release with that info
Announcements are nice for big changes
Broken releases can be removed. Doing a release at a different time is okay for big changes or fixing breaking bugs
You can ask for an announcement in news in the discord

@wally4000 also mentioned he'd like us to have some automated tests to make sure releases are good. I'm not 100% sure how we'd do that and I don't think it would be a blocker, but it's for sure something we do need to work on.

**Comment by sharkwouter** [comment]
Now every repo adds a line to `$PSPDEV/build.txt` :D

**Comment by sharkwouter** [comment]
Okay, we have automatic releases at the beginning of each month now: https://github.com/pspdev/pspdev/pull/80

For the releases we use a date based format. At the moment if you need to do a second release on a day you either have to do so manually or throw away the old release. I do think this is probably rare enough that we don't need a specific solution for this.

Maybe just manually making a v20240619-2 release is the solution here.

**Comment by sharkwouter** [comment, marked as answer]
@fjtrujy we now also have automatically generated changelogs on releases, as seen here: https://github.com/pspdev/pspdev/releases/tag/v20241001

Do you think anything else is needed for this discussion topic?

**Reply by fjtrujy** [reply]
> @fjtrujy we now also have automatically generated changelogs on releases, as seen here: https://github.com/pspdev/pspdev/releases/tag/v20241001
> 
> Do you think anything else is needed for this discussion topic?

Amazing! I think that it is done, isn't?

**Reply by sharkwouter** [reply]
Yeah, I think so :smile:
